tt_eager/tt_dnn/op_library/moreh_helper_functions.hpp

```diff
@@ -126,6 +126,138 @@ struct CircularBufferArg {
 
+
+struct CallbackArgMap {
+    std::map<uint32_t, uint32_t> input;
+    std::map<uint32_t, uint32_t> optional_input;
+    std::map<uint32_t, uint32_t> output;
+};
+
+using Tensors = std::vector<Tensor>;
+using OptionalConstTensors = std::vector<std::optional<const Tensor>>;
+
+// To use this function, the arguments in the reader kernel must always be sorted in the order of input followed by
+// optional_input. Furthermore, input and output tensors must always start from the 0th argument.
+template <typename OutputTensors = Tensors>
+const std::function<void(const void *, Program &, const Tensors &, const OptionalConstTensors &, const OutputTensors &)>
+create_override_runtime_arguments_callback(
+    KernelHandle reader_kernel_id, KernelHandle writer_kernel_id, uint32_t num_cores, uint32_t core_h) {
+    return [reader_kernel_id = reader_kernel_id, writer_kernel_id = writer_kernel_id, num_cores, core_h](
+               const void *operation,
+               Program &program,
+               const Tensors &input_tensors,
+               const OptionalConstTensors &optional_input_tensors,
+               const OutputTensors &output_tensors) -> void {
+        for (uint32_t icore = 0; icore < num_cores; icore++) {
+            CoreCoord core = {icore / core_h, icore % core_h};
+
+            // readers
+            {
+                uint32_t rt_idx = 0;
+                auto &runtime_args = GetRuntimeArgs(program, reader_kernel_id, core);
+                for (uint32_t idx = 0; idx < input_tensors.size(); idx++) {
+                    runtime_args[rt_idx++] = input_tensors.at(idx).buffer()->address();
+                }
+                for (uint32_t idx = 0; idx < optional_input_tensors.size(); idx++) {
+                    auto optional_input_tensor = optional_input_tensors.at(idx);
+                    runtime_args[rt_idx++] =
+                        optional_input_tensor.has_value() ? optional_input_tensor.value().buffer()->address() : 0;
+                }
+            }
+
+            // writer
+            {
+                auto &runtime_args = GetRuntimeArgs(program, writer_kernel_id, core);
+                for (uint32_t idx = 0; idx < output_tensors.size(); idx++) {
+                    runtime_args[idx] = output_tensors.at(idx).buffer()->address();
+                }
+            }
+        }
+    };
+}
+
+// Using this structure is not recommended because directly setting the callback argument map doesn't significantly
+// reduce the amount of code.
+template <typename OutputTensors = Tensors>
+const std::function<void(const void *, Program &, const Tensors &, const OptionalConstTensors &, const OutputTensors &)>
+create_override_runtime_arguments_callback(
+    KernelHandle reader_kernel_id,
+    KernelHandle writer_kernel_id,
+    uint32_t num_cores,
+    uint32_t core_h,
+    CallbackArgMap &arg_map) {
+    return [reader_kernel_id = reader_kernel_id, writer_kernel_id = writer_kernel_id, arg_map, num_cores, core_h](
```

Comment:
I think it would be better to use value capture for arg_map instead of reference capture to avoid validity issues.

Reply:
```C++
    CallbackArgMap &arg_map)  -> CallbackArgMap arg_map)
```
Edited from reference to copy.